File: c++/p101-p200/leetcode199.cpp
// <!-- encoding UTF-8 --!>
/*****************************************************************************
*                     ----Stay Hungry Stay Foolish----                       *
*   @author :   Shen                                                         *
*   @name   :   Leetcode 199                                                 *
*****************************************************************************/

#include <bits/stdc++.h>
using namespace std;
typedef long long int64;
template<class T>inline bool updateMin(T& a, T b){ return a > b ? a = b, 1: 0; }
template<class T>inline bool updateMax(T& a, T b){ return a < b ? a = b, 1: 0; }

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    void dfs(TreeNode* root, vector<int>& res, int dep)
    {
        if (!root) return;
        if (dep >= res.size()) res.push_back(root -> val);
        dfs(root -> right, res, dep + 1);
        dfs(root -> left, res, dep + 1);
    }
public:
    vector<int> rightSideView(TreeNode* root)
    {
        vector<int> res;
        dfs(root, res, 0);
        return res;
    }
};

void test() {}

int main() { test(); return 0; }

File: ruby/leetcode199.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer[]}

def right_side_view(root)
    return [] if not root
    right = right_side_view(root.right)
    left = right_side_view(root.left)
    return [root.val] + right + left.drop(right.length)
end